File: core/src/commonMain/kotlin/work/socialhub/kmastodon/entity/share/Link.java
package work.socialhub.kmastodon.entity.share;

import net.socialhub.http.HttpResponse;
import net.socialhub.logger.Logger;

import java.util.Arrays;

/**
 * @author uakihir0
 */
public class Link {

    private static final Logger logger = Logger.getLogger(Link.class);
    private static final String LINK = "link";

    private String nextUrl;
    private String prevUrl;

    private String nextMaxId;
    private String prevMinId;

    public static Link of(HttpResponse response) {
        if (response == null) {
            return null;
        }

        try {
            String header = response.getResponseHeader(LINK);
            if (header == null || header.isEmpty()) {
                return null;
            }

            // Next と Prev で分割
            String[] parts = header.split(",\\s?");

            Link link = new Link();
            for (String part : parts) {

                // リンクと rel で分割
                String[] elements = part.split(";\\s?");
                if (elements.length == 2) {

                    // Next
                    if (elements[1].contains("next")) {
                        String url = extractUrl(elements[0]);
                        if (!url.isEmpty()) {
                            link.setNextUrl(url);
                            link.setNextMaxId(extractValue(url, "max_id"));
                        }
                    }

                    // Prev
                    else if (elements[1].contains("prev")) {
                        String url = extractUrl(elements[0]);
                        if (!url.isEmpty()) {
                            link.setPrevUrl(url);
                            link.setPrevMinId(extractValue(url, "min_id"));
                        }
                    }
                }
            }
            return link;

        } catch (Exception e) {
            logger.debug(e.getMessage());
            return null;
        }
    }

    private static String extractUrl(String element) {
        return element.replaceAll("^\\s?<(.+)>\\s?$", "$1");
    }

    private static String extractValue(String url, String key) {
        String[] params = url.split("\\?")[1].split("&");

        return Arrays.stream(params)
                .map(e -> e.split("="))
                .filter(e -> e[0].equalsIgnoreCase(key))
                .map(e -> e[1])
                .findFirst()
                .orElse(null);
    }

    public String getNextUrl() {
        return nextUrl;
    }

    public void setNextUrl(String nextUrl) {
        this.nextUrl = nextUrl;
    }

    public String getPrevUrl() {
        return prevUrl;
    }

    public void setPrevUrl(String prevUrl) {
        this.prevUrl = prevUrl;
    }

    public String getNextMaxId() {
        return nextMaxId;
    }

    public void setNextMaxId(String nextMaxId) {
        this.nextMaxId = nextMaxId;
    }

    public String getPrevMinId() {
        return prevMinId;
    }

    public void setPrevMinId(String prevMinId) {
        this.prevMinId = prevMinId;
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/entity/share/Link.kt
package work.socialhub.kmastodon.entity.share

import work.socialhub.khttpclient.HttpResponse
import kotlin.js.JsExport

@JsExport
class Link {
    var nextUrl: String? = null
    var prevUrl: String? = null

    var nextMaxId: String? = null
    var prevMinId: String? = null

    companion object {
        private const val LINK = "link"

        fun of(
            @Suppress("NON_EXPORTABLE_TYPE")
            response: HttpResponse
        ): Link? {
            try {
                val header = response.headers[LINK]
                if (header.isNullOrEmpty()) {
                    return null
                }

                // Next と Prev で分割
                val parts = header[0]
                    .split(",\\s?".toRegex())
                    .dropLastWhile { it.isEmpty() }
                    .toTypedArray()

                val link = Link()
                for (part in parts) {

                    // リンクと rel で分割
                    val elements = part
                        .split(";\\s?".toRegex())
                        .dropLastWhile { it.isEmpty() }
                        .toTypedArray()

                    if (elements.size == 2) {

                        if (elements[1].contains("next")) {

                            // Next
                            val url = extractUrl(elements[0])
                            if (url.isNotEmpty()) {
                                link.nextUrl = url
                                link.nextMaxId = extractValue(url, "max_id")
                            }

                        } else if (elements[1].contains("prev")) {

                            // Prev
                            val url = extractUrl(elements[0])
                            if (url.isNotEmpty()) {
                                link.prevUrl = url
                                link.prevMinId = extractValue(url, "min_id")
                            }
                        }
                    }
                }
                return link

            } catch (e: Exception) {
                return null
            }
        }

        private fun extractUrl(element: String): String {
            return element.replace("^\\s?<(.+)>\\s?$".toRegex(), "$1")
        }

        private fun extractValue(
            url: String,
            key: String,
        ): String? {
            val params =
                url.split("\\?".toRegex())
                    .dropLastWhile { it.isEmpty() }
                    .toTypedArray()[1]
                    .split("&")
                    .dropLastWhile { it.isEmpty() }
                    .toTypedArray()

            return params
                .map { p ->
                    p.split("=")
                        .dropLastWhile { it.isEmpty() }
                        .toTypedArray()
                }
                .filter { it[0].equals(key, ignoreCase = true) }
                .map { it[1] }
                .firstOrNull()
        }
    }
}
